src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -34,91 +34,197 @@ def self.compile_ast(code, compile_options, boltlib_path)
+            'transport' => 'redacted',
+            'transports' => {
+              'redacted' => 'redacted'
+            }
+          }
+          bolt_inv = Bolt::ApplyInventory.new({fake_config})
+          Puppet.override(bolt_inventory: bolt_inv) do
+            Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
             # This compiler has been configured with a node containing
             # the requested environment, facts, and variables, and is used
             # to compile a catalog in that context from the supplied AST.
-            pal.with_catalog_compiler do |compiler|
-              # We have to parse the AST inside the compiler block, because it
-              # initializes the necessary type loaders for us.
-              ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+            pal.with_catalog_compiler(**variables) do |compiler|
+              # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
 
+              ast = build_program(code)
               compiler.evaluate(ast)
+              compiler.evaluate_ast_node
               compiler.compile_additions
               compiler.catalog_data_hash
             end
           end
         end
       end
-      private_class_method :compile_ast
+      private_class_method :compile_for_environment_plan
+
+      def self.compile_for_environment(code, compile_options, boltlib_path)
+        # When we do not need to load bolt we assume there are no Bolt types to resolve in
+        # AST compilation.
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                    envpath: Puppet[:environmentpath],
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+
+      def self.with_puppet_settings(overrides = {})
```

Comment:
This function seems like the most dangerous part of the functionality for compiling for projects. In it we basically attempt to clear all settings for puppet so that the compile in the project context isn't polluted by settings fetched automatically by puppet.

This was taken almost wholesale from what we do in Bolt, but we need to discuss what impact this has on the JRuby environment and if we can even use this function or something like it.

Reply:
Unlike in Bolt, where something runs once and then is done, this loaded Puppet instance will continue to be used after the project compile is done, and in that case, we need all of the original settings.

Maybe there is a way you can leverage the Settings object we define here in Puppetserver, which has a concept of "local" settings overrides: https://github.com/puppetlabs/puppetserver/blob/2f9122bb2198f2b1e7b0ee638c33da8766f26a8b/src/ruby/puppetserver-lib/puppet/server/settings.rb#L3 It's designed for multithreaded, but might be usable here too with some extension, e.g. a way to create the "local" settings with more than one new value at once. And then I think you can just call `clear_local_settings` when you're done to go back to our main global ones. Worth looking into anyway, I didn't think about it too hard, so hopefully this isn't super off-base.